### PR TITLE
feat: implement bare metal deployer

### PR DESCRIPTION
**Pull request by zyy17**

### CODE REVIEW

File: pkg/deployer/baremetal/config.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+)
+
+// Config is the desired state of a GreptimeDB cluster on bare metal.
+type Config struct {
+	Cluster *Cluster `yaml:"cluster"`
+	Etcd    *Etcd    `yaml:"etcd"`
+}
+
+type Cluster struct {
+	Name     string    `yaml:"name"`
```
Comment (daviderli614) [inline pkg/deployer/baremetal/config.go]: question: Cluster name double definition, we use `clusterName = args[0]` ?

**Comment by daviderli614** [comment]
issue: 
Run `./bin/gtctl cluster create mycluster --bare-metal --config examples/bare-metal/cluster-with-local-artifacts.yaml`
The following error:
```
Installing etcd cluster failed
fork/exec /xxx/.gtctl/artifacts/greptime/latest/bin: permission denied
```

My cluster-with-local-artifacts.yaml:
```
cluster:
   name: mycluster # name of the cluster
   artifact:
     local: "/xxx/.gtctl/artifacts/etcd/v3.5.7/bin"
   frontend: {}
   datanode:
     replicas: 3
     rpcAddr: 0.0.0.0:14100
     mysqlAddr: 0.0.0.0:14200
     httpAddr: 0.0.0.0:14300
   meta:
     storeAddr: 127.0.0.1:2379
     serverAddr: 0.0.0.0:3002
     httpAddr: 0.0.0.0:14001

etcd:
   artifact:
     local: "/xxx/.gtctl/artifacts/greptime/latest/bin"
```

File: pkg/deployer/baremetal/config.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func (artifact *Artifact) validate() error {
+	if artifact.Version == "" && artifact.Local == "" {
+		return fmt.Errorf("empty artifact")
+	}
```
Comment (daviderli614) [inline pkg/deployer/baremetal/config.go:188]: thought: `if artifact.Version != "" && artifact.Local != ""`, which binary will be use to create the cluster?

File: pkg/deployer/baremetal/config.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+func (cluster *Cluster) validate() error {
+	if cluster.Artifact == nil {
+		return fmt.Errorf("cluster artifact field is nil")
+	}
+	if err := cluster.Artifact.validate(); err != nil {
+		return err
+	}
+
+	if cluster.Frontend == nil {
```
Comment (daviderli614) [inline pkg/deployer/baremetal/config.go:103]: question: The `cluster.Frontend` always nil in `defaultConfig` function.

File: pkg/deployer/baremetal/config.go (hunk 1 above)
Comment (zyy17) [reply pkg/deployer/baremetal/config.go]: @daviderli614, you're right, the `name` is unnecessary, I will remove it.

File: pkg/deployer/baremetal/config.go (hunk 2 above)
Comment (zyy17) [reply pkg/deployer/baremetal/config.go:188]: @daviderli614 it will use the `Local` binary.

**Comment by zyy17** [comment]
@daviderli614 The reason of why you can't use local binary is you have specify the full binary path instead of directory, for example: `$HOME/.gtctl/artifacts/greptime/v0.2.0-nightly-20230403/bin/greptime`.

File: pkg/deployer/baremetal/config.go (hunk 3 above)
Comment (zyy17) [reply pkg/deployer/baremetal/config.go:103]: @daviderli614 It's [not nil](https://github.com/zyy17/gtctl/blob/feat/bare-metal-deployer/pkg/deployer/baremetal/config.go#L224).

**Comment by zyy17** [comment]
@daviderli614 the bug is fixed https://github.com/zyy17/gtctl/commit/6b66f092fd9bce2a0508ba21ec5f574855d5b8b9

**Review by daviderli614 — APPROVED** [review]
LGTM

**Pull request merged**